crates/bevy_ecs/src/reflect/world.rs

```diff
@@ -0,0 +1,152 @@
+        &mut self,
+        entity: Entity,
+        component_id: ComponentId,
+    ) -> Option<&mut T::Dyn> {
+        let type_id = self.component_type_id(component_id)?;
+
+        let type_data: Box<T> = ({
+            let type_registry = self.get_resource::<AppTypeRegistry>()?;
+            let type_registry = type_registry.read();
+            type_registry.get_type_data::<T>(type_id)?.clone_type_data()
+        })
+        .downcast::<T>()
+        .ok()?;
+        let dyn_obj = self.get_dyn_reflect_mut_by_id(entity, component_id)?;
+        type_data.get_mut(dyn_obj)
+    }
+
+    /// Retrieves an immutable `dyn Reflect` reference to the given entity's Component of the given [`ComponentId`]
+    pub fn get_dyn_reflect_by_id(
+        &self,
+        entity: Entity,
+        component_id: ComponentId,
+    ) -> Option<&dyn Reflect> {
+        let type_id = self.component_type_id(component_id)?;
+        let component_ptr = self.get_by_id(entity, component_id)?;
+
+        let type_registry = self.get_resource::<AppTypeRegistry>()?;
+        let type_registry = type_registry.read();
+        let from_ptr = type_registry.get_type_data::<ReflectFromPtr>(type_id)?;
+        // SAFETY: type_id is correct
+        Some(unsafe { from_ptr.as_reflect(component_ptr) })
+    }
+
+    /// Retrieves an mutable `dyn Reflect` reference to the given entity's Component of the given [`ComponentId`]
+    pub fn get_dyn_reflect_mut_by_id(
+        &mut self,
+        entity: Entity,
+        component_id: ComponentId,
+    ) -> Option<&mut dyn Reflect> {
+        let type_id = self.component_type_id(component_id)?;
+        let reflect_component_ptr = {
+            let type_registry = self.get_resource::<AppTypeRegistry>()?;
+            let type_registry = type_registry.read();
+            type_registry
+                .get_type_data::<ReflectFromPtr>(type_id)?
+                .clone()
+        };
+
+        let mut_ptr = self.get_mut_by_id(entity, component_id)?;
+        Some(
+            mut_ptr
+                .map_unchanged(|p| {
+                    // SAFETY: type_id is correct
+                    unsafe { reflect_component_ptr.as_reflect_mut(p) }
+                })
+                .value,
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use bevy_reflect::{reflect_trait, Reflect};
+
+    use crate::prelude::AppTypeRegistry;
+    use crate::{self as bevy_ecs, component::Component, world::World};
+
+    #[reflect_trait]
+    trait DoThing {
+        fn do_thing(&self) -> String;
+        fn mut_do_thing(&mut self);
+    }
+
+    #[derive(Component, Reflect)]
+    #[reflect(DoThing)]
+    struct ComponentA(String);
+
+    impl DoThing for ComponentA {
+        fn do_thing(&self) -> String {
+            format!("ComponentA {} do_thing!", self.0)
+        }
+
+        fn mut_do_thing(&mut self) {
+            self.0 = "value changed".to_string();
+        }
+    }
+
+    #[test]
+    fn dyn_reflect() {
```

Comment:
I know this is mainly a compile test, but it could be good to have some sort of assertion in here to make sure we're reading/mutating the correct entity (and not some other component with the same `ComponentId`).